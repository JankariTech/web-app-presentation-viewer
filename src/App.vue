<template>
  <div
    id="presentation-viewer-main"
    class="presentation-viewer oc-flex"
    :class="{ 'dark-mode': isDarkMode }"
  >
    <div ref="revealContainer" class="reveal">
      <div id="slideContainer" ref="slideContainer" class="slides">
        <section
          :data-markdown="url"
          :data-separator="dataSeparator"
          :data-separator-vertical="dataSeparatorVertical"
        ></section>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { computed, onMounted, onBeforeUnmount, ref, unref, watch } from 'vue'
import {
  useThemeStore,
  useAppDefaults,
  useAppFileHandling,
  useClientService,
  useAppsStore,
  useConfigStore
} from '@ownclouders/web-pkg'
import { Resource } from '@ownclouders/web-client/src'
import Reveal from 'reveal.js'
import RevealMarkdown from 'reveal.js/plugin/markdown/markdown'
import RevealHighlight from 'reveal.js/plugin/highlight/highlight'

import 'reveal.js/dist/reveal.css'
import 'reveal.js/plugin/highlight/monokai.css'
import 'reveal.js/dist/theme/white.css'

import { getMediaMimeTypes } from './helpers/mediaMimeTypes'
import { id as appId } from '../public/manifest.json'

const themeStore = useThemeStore()
const clientService = useClientService()
const { webdav } = clientService
const { loadFolderForFileContext, currentFileContext, activeFiles } = useAppDefaults({
  applicationId: appId
})
const { getUrlForResource, revokeUrl } = useAppFileHandling({ clientService })
const appsStore = useAppsStore()
const { serverUrl } = useConfigStore()

const isDarkMode = ref(themeStore.currentTheme.isDark)
const slideContainer = ref<HTMLElement>()
const revealContainer = ref<HTMLElement>()
const mediaUrls = ref<string[]>([])

const mediaBasePath = `${serverUrl}local/`
const dataSeparator = '\r?\n---\r?\n'
const dataSeparatorVertical = '\r?\n--\r?\n'

let reveal: Reveal.Api

defineProps({
  url: {
    type: String,
    required: true
  }
})

watch(
  () => themeStore.currentTheme.isDark,
  (isDark) => {
    isDarkMode.value = isDark
  }
)

// LIFECYCLE HOOKS
onMounted(async () => {
  await loadFolderForFileContext(unref(currentFileContext))

  reveal = new Reveal(unref(revealContainer), {
    plugins: [RevealMarkdown, RevealHighlight]
  })

  await reveal.initialize({
    controls: true,
    progress: true,
    history: true,
    center: true,
    controlsLayout: 'edges',
    markdown: {
      baseUrl: mediaBasePath
    }
  })

  reveal.on('ready', async () => {
    if (unref(slideContainer) === undefined) {
      return
    }
    const imgElements = (unref(slideContainer) as HTMLElement).getElementsByTagName('img')
    const localImgElements = filterLocalImgElements(imgElements)
    await updateImageUrls(localImgElements)
  })
})
onBeforeUnmount(() => {
  unref(mediaUrls).forEach((url) => {
    revokeUrl(url)
  })
})

// COMPUTED
const mediaMimeTypes = computed(() =>
  getMediaMimeTypes(appsStore.externalAppConfig['preview']?.mimeTypes)
)
const mediaFiles = computed<Resource[]>(() => {
  if (!unref(activeFiles)) {
    return []
  }

  return unref(activeFiles).filter((file: Resource) => {
    if (!file.mimeType) {
      return false
    }
    return unref(mediaMimeTypes).includes(file.mimeType.toLowerCase())
  })
})

// METHODS
function filterLocalImgElements(
  imgElements: HTMLCollectionOf<HTMLImageElement>
): HTMLImageElement[] {
  const localImgElements: HTMLImageElement[] = []
  for (const el of imgElements) {
    if (el.src.startsWith(mediaBasePath)) {
      localImgElements.push(el)
    }
  }
  return localImgElements
}
async function updateImageUrls(localImgElements: HTMLImageElement[]) {
  for (const el of localImgElements) {
    // trim 'mediaBasePath'
    // remove leading '.' and '/'
    const srcPath = el.src.replace(mediaBasePath, '').replace(/$\.\//, '').replace(/$\//, '')

    const blobUrl = await parseImageUrl(srcPath)
    if (blobUrl) {
      el.src = blobUrl
      mediaUrls.value.push(blobUrl)
    }
  }
}
async function parseImageUrl(name: string) {
  let file: Resource
  if (name.split('/').length > 1) {
    file = await getSubMediaFile(name)
  } else {
    file = getMediaFile(name)
  }
  const url = await getUrlForResource(unref(currentFileContext).space, file)
  // reload the active files
  // TODO: implement caching
  await loadFolderForFileContext(unref(currentFileContext))
  return getBlobUrl(url)
}
function getMediaFile(name: string): Resource {
  return unref(mediaFiles).find((file: Resource) => file.name === name)
}
async function getSubMediaFile(path: string): Promise<Resource> {
  const parents = dirname(path).split('/')
  const filename = basename(path)

  let currentFiles = unref(activeFiles)
  for (const parent of parents) {
    const file = findFile(parent, currentFiles)
    const { children } = await webdav.listFiles(unref(currentFileContext).space, {
      path: file.path,
      fileId: file.fileId
    })

    currentFiles = children

    if (parent === parents[parents.length - 1]) {
      return findFile(filename, children)
    }
  }
}
function findFile(name: string, filesList: Resource[]): Resource {
  return filesList.find((file: Resource) => file.name === name)
}
async function getBlobUrl(url: string) {
  const data = await fetch(url).then(async (res) => await res.blob())
  const blob = new Blob([data], { type: data.type })
  return URL.createObjectURL(blob)
}
function dirname(path: string) {
  return path.split('/').slice(0, -1).join('/')
}
function basename(path: string) {
  return path.split('/').reverse()[0]
}
</script>

<style lang="scss">
.dark-mode {
  .reveal {
    color: var(--oc-color-text-default) !important;

    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      color: var(--oc-color-text-default) !important;
    }
  }
}

#slideContainer {
  section {
    display: flex !important;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;

    > * {
      margin-top: 0;
    }

    p:has(img) {
      display: flex;
      flex-direction: column;
      align-items: center;
      overflow: hidden;

      img {
        width: auto;
        height: auto;
      }
    }
  }
}
</style>
