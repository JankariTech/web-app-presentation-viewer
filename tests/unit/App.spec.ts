import { shallowMount, flushPromises } from '@vue/test-utils'
import App from '../../src/App.vue'

vi.mock('@ownclouders/web-pkg', () => ({
  useAppDefaults: vi.fn().mockImplementation(() => ({
    loadFolderForFileContext: vi.fn(),
    currentFileContext: {},
    activeFiles: []
  })),
  useAppFileHandling: vi.fn().mockImplementation(() => ({
    getUrlForResource: vi.fn(),
    revokeUrl: vi.fn()
  })),
  useThemeStore: vi.fn().mockImplementation(() => ({
    currentTheme: { isDark: false }
  })),
  useClientService: vi.fn(),
  useAppsStore: vi.fn(),
  useConfigStore: vi.fn().mockImplementation(() => ({
    serverUrl: 'https://localhost:9200'
  }))
}))
// vi.mock('reveal.js', async (importOriginal) => {
//   const actual = await importOriginal()
//   return {
//     ...actual,
//     initialize: vi.fn()
//   }
// })

describe('App component', () => {
  it('mount component', async () => {
    const vm = getWrapper()
    await flushPromises()
    console.log(vm.html())
    // expect(vm.html()).toMatchSnapshot()
  })
})

function getWrapper() {
  const blobUrl = URL.createObjectURL(new Blob(['## Slide 1']))
  return shallowMount(App, {
    propsData: { url: blobUrl }
  })
}
