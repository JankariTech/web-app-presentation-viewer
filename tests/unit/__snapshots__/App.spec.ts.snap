// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > render markdown slides 1`] = `
"<div id="presentation-viewer-main" class="presentation-viewer">
  <spy-stub></spy-stub>
  <div role="application" data-transition-speed="default" data-background-transition="fade" class="reveal slide center focused has-horizontal-slides">
    <div id="slideContainer" class="slides no-transition" style="width: 960px; height: 700px; left: 50%; top: 50%; bottom: auto; right: auto; transform: translate(-50%, -50%) scale(0.2);">
      <section data-markdown="" data-markdown-parsed="true" style="top: 350px; display: block;" class="present">
        <h3 id="slide-1">Slide 1</h3>
        <p>Lorem <strong>Bold</strong> <em>Italic</em> <del>strike through</del></p>
        <blockquote>
          <p>Quote statement</p>
        </blockquote>
        <ul>
          <li><input disabled="" type="checkbox"> check box</li>
        </ul>
        <ul>
          <li><input disabled="" type="checkbox"> check box</li>
        </ul>
      </section>
      <section data-markdown="" data-markdown-parsed="true" style="top: 350px; display: block;" hidden="" aria-hidden="true" class="future">
        <h3 id="slide-2">Slide 2</h3>
        <p>Unordered list:</p>
        <ul>
          <li>item 1</li>
        </ul>
        <ul>
          <li>item 2</li>
        </ul>
        <p>Ordered list:</p>
        <ol>
          <li>item 1</li>
          <li>item 2</li>
        </ol>
      </section>
      <section data-markdown="" data-markdown-parsed="true" style="top: 350px; display: block;" hidden="" aria-hidden="true" class="future">
        <h3 id="external-image">External Image</h3>
        <p><img src="https://external:9200/cat.jpg" alt="cat"></p>
      </section>
      <section data-markdown="" data-markdown-parsed="true" style="top: 350px; display: none;" hidden="" aria-hidden="true" class="future">
        <h3 id="uploaded-image">Uploaded Image</h3>
        <p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAcIAAAHCCAYAAAB8GMlFAAAAAXNSR0IArs4c6QAAIABJREFUeF==" alt=""></p>
      </section>
      <section data-markdown="" data-markdown-parsed="true" style="top: 350px; display: none;" hidden="" aria-hidden="true" class="future">
        <h3 id="occis-image">ocCIS Image</h3>
        <p><img src="blob:nodedata:0295bafb-5976-468a-a263-685a8872cb96" alt="cool"></p>
        <p><img src="blob:nodedata:0295bafb-5976-468a-a263-685a8872cb96" alt="cool"></p>
        <p><img src="./non-existing-image.png" alt="non-existing"></p>
        <p><img src="blob:nodedata:0295bafb-5976-468a-a263-685a8872cb96" alt="sub-folder-image"></p>
        <p><img src="./non-existing/another-cool.jpg" alt="sub-folder-image"></p>
      </section>
      <section data-markdown="" data-markdown-parsed="true" style="top: 350px; display: none;" hidden="" aria-hidden="true" class="future">
        <h3 id="code-block">Code block</h3>
        <pre class="code-wrapper"><code data-highlighted="yes" class="hljs language-clean"><span class="hljs-keyword">code</span> block</code></pre>
      </section>
    </div>
    <div class="backgrounds no-transition">
      <div style="display: block;" data-loaded="true" class="slide-background present">
        <div class="slide-background-content"></div>
      </div>
      <div style="display: block;" data-loaded="true" class="slide-background future">
        <div class="slide-background-content"></div>
      </div>
      <div style="display: block;" data-loaded="true" class="slide-background future">
        <div class="slide-background-content"></div>
      </div>
      <div style="display: none;" class="slide-background future">
        <div class="slide-background-content"></div>
      </div>
      <div style="display: none;" class="slide-background future">
        <div class="slide-background-content"></div>
      </div>
      <div style="display: none;" class="slide-background future">
        <div class="slide-background-content"></div>
      </div>
    </div>
    <div class="slide-number" style="display: none;"></div>
    <aside class="controls" style="display: block;" data-controls-layout="edges" data-controls-back-arrows="faded"><button aria-label="previous slide" class="navigate-left" disabled="disabled">
        <div class="controls-arrow"></div>
      </button>
      <button aria-label="next slide" class="navigate-right enabled highlight">
        <div class="controls-arrow"></div>
      </button>
      <button aria-label="above slide" class="navigate-up" disabled="disabled">
        <div class="controls-arrow"></div>
      </button>
      <button aria-label="below slide" disabled="disabled" class="navigate-down">
        <div class="controls-arrow"></div>
      </button>
    </aside>
    <div class="progress" style="display: block;"><span style="transform: scaleX(0);"></span></div>
    <div class="speaker-notes" data-prevent-swipe="" tabindex="0"></div>
    <div class="pause-overlay"><button class="resume-button">Resume presentation</button></div>
    <div style="position: absolute; height: 1px; width: 1px; overflow: hidden;" class="aria-status" aria-live="polite" aria-atomic="true"></div>
  </div>
</div>"
`;
